axum-extra/src/multipart_builder.rs

```diff
@@ -0,0 +1,330 @@
+    ///
+    /// This function will return an error if the provided MIME type is not valid.
+    ///
+    /// # Examples
+    ///
+    /// ```rust
+    /// use axum_extra::multipart_builder::{MultipartForm, Part, TransferEncoding};
+    ///
+    /// // create a form with a single part that has a field with a name of "part_name",
+    /// // with a MIME type of "application/json", and the supplied contents. This part will not have an associated filename, but will be sent as binary, and does not
+    /// // need to be valid UTF-8.
+    /// let parts: Vec<Part> = vec![Part::raw_part("part_name", "application/json", vec![0x68, 0x68, 0x20, 0x6d, 0x6f, 0x6d], None, TransferEncoding::Binary).expect("MIME type must be valid")];
+    /// let form = MultipartForm::with_parts(parts);
+    /// ```
+    pub fn raw_part(
+        name: &str,
+        mime_type: &str,
+        contents: Vec<u8>,
+        filename: Option<&str>,
+        encoding: TransferEncoding,
+    ) -> Result<Self, &'static str> {
+        let mime_type = mime_type.parse().map_err(|_| "Invalid MIME type")?;
+        Ok(Self {
+            name: name.to_owned(),
+            filename: filename.map(|f| f.to_owned()),
+            mime_type,
+            contents,
+            encoding,
+        })
+    }
+
+    /// Serialize this part into a chunk that can be easily inserted into a larger form
+    pub(super) fn serialize(&self) -> Vec<u8> {
+        // A part is serialized in this general format:
+        // // the filename is optional
+        // Content-Disposition: form-data; name="FIELD_NAME"; filename="FILENAME"\r\n
+        // // the mime type (not strictly required by the spec, but always sent here)
+        // Content-Type: mime/type\r\n
+        // // if the part does not conform to the rest of the request's encoding,
+        // // this is specified
+        // Content-Transfer-Encoding: "ENCODING"\r\n
+        // // a blank line, then the contents of the file start
+        // \r\n
+        // CONTENTS\r\n
+
+        // Format what we can as a string, then handle the rest at a byte level
+        let mut serialized_part = format!("Content-Disposition: form-data; name=\"{}\"", self.name);
+        // specify a filename if one was set
+        if let Some(filename) = &self.filename {
+            serialized_part += &format!("; filename=\"{}\"", filename);
+        }
+        serialized_part += "\r\n";
+        // specify the MIME type
+        serialized_part += &format!("Content-Type: {}\r\n", self.mime_type);
+        // if an encoding was set, add that
+        // determine what encoding to label the body of the field with
+        let encoding: Option<&str> = match self.encoding {
+            TransferEncoding::TextUTF8 => None,
+            TransferEncoding::Binary => Some("binary"),
+        };
+        if let Some(encoding) = encoding {
+            serialized_part += &format!("Content-Transfer-Encoding: {}\r\n", encoding);
+        }
```

Comment:
Are you sure that this is a thing? I googled and found `Content-Encoding` and `Transfer-Encoding` docs, but the only mention of `Content-Transfer-Encoding` was in relation to some old RFC that has been replaced since (AFAIU).

Reply:
So I actually implemented most of this using RFC 2388 (whoops), which has been superceded by RFC 7578, which deprecates that field. I updated the comments and code to follow RFC 5758